### ISSUE TITLE
Issue with clearOnRefresh

### ISSUE THREAD

**Issue by mdenoronha** [formio/formio.js#2227]
This issue is with clearOnRefresh. 

While it's working on some fields, other fields do not refresh as intended. The intended functionality is when the first field in the row is changed, the second should refresh with new options from a URL. While the new options are retrieved and added, the previously selected option persists (with its label changed to its value).

![ezgif-4-bc297df538db](https://user-images.githubusercontent.com/23422862/71484086-7f5bf500-2802-11ea-95c2-25b90197b215.gif)

The issue concerns fields 'webhook_data' and 'is' below
```js
Formio.icons = 'fontawesome';
Formio.createForm(document.getElementById('formio'), {
			components: [
			{
				type: 'select',
				label: 'Trigger',
				key: 'trigger',
				placeholder: 'Select your trigger',
				dataSrc: 'url',
				data: {
					url: '{{host}}{{ url_for('static', filename='json/triggers.json') }}'
				},
				valueProperty: 'webhook',
				template: '{% raw %}<span class="form__trigger-input form__trigger-input--{{ item.input }}"><span>{{ item.trigger }}</span><br><span class="form__trigger-info">{{ item.info }}</span></span>{% endraw %}',
				selectValues: 'results',
				lazyLoad: true
			},
			{
		        label: 'If:',
		        key: 'conditions',
		        type: 'datagrid',
		        input: true,
		        validate: {
		          minLength: 1,
		          maxLength: 6
		        },
		        components: [
		          {
		          	type: 'select',
		          	label: '',
		          	key: 'webhook_data',
		          	placeholder: 'Select your data',
		          	dataSrc: 'url',
		          	data: {
		          	  url: '{{ host }}{% raw %}/webhooks-data/{{ data.trigger }}?format=json{% endraw %}'
		          	},
		          	valueProperty: 'data_value',
		          	template: '{% raw %}<span>{{ item.data_name }}</span>{% endraw %}',
		          	refreshOn: 'trigger',
		          	clearOnRefresh: true,
		          	selectValues: 'response_data',
		          	validate: {
		          	  required: true
		          	}
		          },
		          {
		            label: '',
		            key: 'any',
		            type: 'select',
		            input: true,
		            data: {
		              values: [
		                {
		                  value: 'any',
		                  label: '{% raw %}Any match the following{% endraw %}'
		                },
		                {
		                  value: 'all',
		                  label: '{% raw %}All match the following{% endraw %}'
		                },
		                {
		                  value: 'none',
		                  label: '{% raw %}None of match the following{% endraw %}'
		                }
		              ]
		            },
		            dataSrc: 'values',
		            refreshOn: 'webhook_data',
		            clearOnRefresh: true,
		            template: '{% raw %}<span>{{ item.label }}</span>{% endraw %}',
		            conditional: {
		              json: {
		                "in": [
		                  "multiple",
		                  {
		                	"var": "row.webhook_data"
		                  }
		                ]
		              }
		            }
		          },
		          {
		          	type: 'select',
		          	label: '',
		          	key: 'is',
		          	dataSrc: 'url',
		          	data: {
		          	  url: '{{ host }}{% raw %}/webhooks-data-is/{{ row.webhook_data }}?format=json{% endraw %}'
		          	},
		          	valueProperty: 'value',
		          	template: '{% raw %}<span>{{ item.input_name }}</span>{% endraw %}',
		          	refreshOn: 'webhook_data',
		          	clearOnRefresh: true,
		          	selectValues: 'inputs',
		          	validate: {
		          	  required: true
		          	}
		          },
		          {
		            label: '',
		            key: 'lastName',
		            type: 'textfield',
		            input: true
		          }
		        ]
		      },
			{
		    	type: 'button',
		    	action: 'submit',
		    	label: 'Submit',
		    	theme: 'primary'
		    }
			]
		}).then(function(form) {
			form.on('submit', function(submission) {
			  console.log(submission);
			  return fetch('/', {
			      body: JSON.stringify(submission),
			      headers: {
			        'content-type': 'application/json'
			      },
			      method: 'POST',
			      mode: 'cors',
			    })
			    .then(response => {
			    	console.log('submit')
			    })
			});
		});
```

**Comment by jhen1422** [comment]
@mdenoronha Thank you for the report. Can you please provide a fiddle showcasing the issue?

**Comment by mdenoronha** [comment]
Sure, here you go https://jsfiddle.net/bxr6w35m/2/

**Comment by mdenoronha** [comment]
To replicate the issue, select 'A new paid order is made' from the 'trigger' input, then 'note' from the first column input in the grid, then 'is not exactly' from the second column input, then change the first column input to 'created at'. The second column will display the issue.

**Comment by jhen1422** [comment]
@mdenoronha
We have several examples where 'Clear On Refresh' is working properly. Can you verify if the 'Refresh Options On' field is set to the first dropdown in the grid?

Also, this could be occurring because the values for Dropdown 2 share the same values when different values from Dropdown 1 are selected. The 'is not exactly' value looks like it displays  when both 'A new paid order is made' and 'note' values are selected from dropdown 1. Since the values are the same, the field is not registering a change.

**Comment by mdenoronha** [comment]
> We have several examples where 'Clear On Refresh' is working properly. Can you verify if the 'Refresh Options On' field is set to the first dropdown in the grid?

Yep, the `refreshOn ` field is set to `webhook_data ` which is the first grid dropdown.

> Also, this could be occurring because the values for Dropdown 2 share the same values when different values from Dropdown 1 are selected. The 'is not exactly' value looks like it displays when both 'A new paid order is made' and 'note' values are selected from dropdown 1. Since the values are the same, the field is not registering a change.

The values are different in Grid Dropdown 2 (key: is) when different values in Grid Dropdown 1 (key: webhook_data) is selected. That's working fine. As intended changing from 'Note' to 'Created At' in Grid Dropdown 1 changes the values in Grid Dropdown 2. 

As the gif shows, it's just the previously selected option that stays and is appended to the new options. 'not_exactly'/'Is Not Exactly' is not an option for 'Created At', it was the previously selected option for 'Note'. 

As well, randomly its name switches to its value e.g. 'Is Not Exactly' becomes 'not_exactly'; which was its value.

**Comment by randallknutson** [comment]
This may be caused by it being in a datagrid. It is supposed to work in a datagrid but may be using the wrong value to calculate the logic to run clear on hide. We will add the the to do list to get this resolved.

**Comment by mdenoronha** [comment]
@randallknutson Ok, great thanks a lot.
Is there an easy workaround way to clear a formio form field in the meantime? Simulate the close button maybe?

**Comment by travist** [comment]
Please verify you have this checkbox enabled.

https://monosnap.com/file/jL1NmHLcTkbjAejNgbHeCSrlEDiWZE

**Comment by mdenoronha** [comment]
@travist I didn't use the form builder, what option would that be? `clearOnRefresh `? Because as the example above shows it's set to `True`

**Comment by randallknutson** [comment]
That is the setting (clearOnRefresh). If it shows as true it should work correctly.

**Comment by beltschatsar** [comment]
Same issue for me. Would you accept a PR ?

**Comment by randallknutson** [comment]
We always love PRs. If you create one we can review it and make sure it addresses the issue and get it merged.

**Comment by Sefriol** [comment]
This problem seems to also occur when you put a submission into a form. For me it seems that fields are not properly assigned with a new value and they will keep a value that was previously included in the form.

For me, clearOnHide won't work if:
1. Assign a submission into a form
2. Hide a field with submitted data
3. When shown again fields are still populated and when pressing a submit button, these fields will be sent with the information

This will not happen however if:
1. Assign a submission into a form
2. Do any event on a field that will trigger a component change (change the input as an example, or trigger hider element a second time)

Form example (tested on with latest formio-vue with formiojs 4.8.1). I have yet to test this on plain project.
```json
{
  "components": [
    {
      "label": "Radio",
      "labelPosition": "top",
      "optionsLabelPosition": "right",
      "description": "",
      "tooltip": "",
      "customClass": "",
      "tabindex": "",
      "inline": false,
      "hidden": false,
      "hideLabel": false,
      "autofocus": false,
      "disabled": false,
      "tableView": false,
      "modalEdit": false,
      "values": [
        {
          "label": "True",
          "value": "true",
          "shortcut": ""
        },
        {
          "label": "False",
          "value": "false",
          "shortcut": ""
        }
      ],
      "dataType": "",
      "persistent": true,
      "protected": false,
      "dbIndex": false,
      "encrypted": false,
      "redrawOn": "",
      "clearOnHide": true,
      "customDefaultValue": "",
      "calculateValue": "",
      "allowCalculateOverride": false,
      "validate": {
        "required": false,
        "customMessage": "",
        "custom": "",
        "customPrivate": false,
        "json": "",
        "strictDateValidation": false,
        "unique": false,
        "multiple": false
      },
      "errorLabel": "",
      "key": "radio1",
      "tags": [],
      "conditional": {
        "show": null,
        "when": null,
        "eq": "",
        "json": ""
      },
      "customConditional": "",
      "logic": [],
      "overlay": {
        "style": "",
        "page": "",
        "left": "",
        "top": "",
        "width": "",
        "height": ""
      },
      "type": "radio",
      "input": true,
      "placeholder": "",
      "prefix": "",
      "suffix": "",
      "multiple": false,
      "unique": false,
      "refreshOn": "",
      "widget": null,
      "validateOn": "change",
      "showCharCount": false,
      "showWordCount": false,
      "allowMultipleMasks": false,
      "inputType": "radio",
      "fieldSet": false,
      "id": "ertrth",
      "defaultValue": "",
      "attributes": {},
      "properties": {}
    },
    {
      "label": "data123",
      "labelPosition": "top",
      "placeholder": "",
      "description": "",
      "tooltip": "",
      "prefix": "",
      "suffix": "",
      "widget": {
        "type": "input"
      },
      "inputMask": "",
      "allowMultipleMasks": false,
      "customClass": "",
      "tabindex": "",
      "hidden": false,
      "hideLabel": false,
      "showWordCount": false,
      "showCharCount": false,
      "mask": false,
      "autofocus": false,
      "spellcheck": true,
      "disabled": false,
      "tableView": true,
      "modalEdit": false,
      "multiple": false,
      "persistent": true,
      "inputFormat": "plain",
      "protected": false,
      "dbIndex": false,
      "case": "",
      "encrypted": false,
      "redrawOn": "",
      "clearOnHide": true,
      "customDefaultValue": "",
      "calculateValue": "",
      "allowCalculateOverride": false,
      "validateOn": "change",
      "validate": {
        "required": false,
        "pattern": "",
        "customMessage": "",
        "custom": "",
        "customPrivate": false,
        "json": "",
        "minLength": "",
        "maxLength": "",
        "unique": false,
        "multiple": false,
        "strictDateValidation": false
      },
      "unique": false,
      "errorLabel": "",
      "key": "data123",
      "tags": [],
      "conditional": {
        "show": true,
        "when": "radio1",
        "eq": "true",
        "json": ""
      },
      "customConditional": "",
      "logic": [],
      "overlay": {
        "style": "",
        "page": "",
        "left": "",
        "top": "",
        "width": "",
        "height": ""
      },
      "type": "textfield",
      "input": true,
      "refreshOn": "",
      "inputType": "text",
      "id": "eco2tmf",
      "defaultValue": "",
      "attributes": {},
      "properties": {}
    },
    {
      "type": "button",
      "label": "Submit",
      "key": "submit",
      "size": "md",
      "block": false,
      "action": "submit",
      "disableOnInvalid": true,
      "theme": "primary",
      "input": true,
      "placeholder": "",
      "prefix": "",
      "customClass": "",
      "suffix": "",
      "multiple": false,
      "defaultValue": null,
      "protected": false,
      "unique": false,
      "persistent": false,
      "hidden": false,
      "clearOnHide": true,
      "refreshOn": "",
      "redrawOn": "",
      "tableView": false,
      "modalEdit": false,
      "labelPosition": "top",
      "description": "",
      "errorLabel": "",
      "tooltip": "",
      "hideLabel": false,
      "tabindex": "",
      "disabled": false,
      "autofocus": false,
      "dbIndex": false,
      "customDefaultValue": "",
      "calculateValue": "",
      "widget": {
        "type": "input"
      },
      "validateOn": "change",
      "validate": {
        "required": false,
        "custom": "",
        "customPrivate": false,
        "strictDateValidation": false,
        "unique": false,
        "multiple": false
      },
      "conditional": {
        "show": null,
        "when": null,
        "eq": ""
      },
      "overlay": {
        "style": "",
        "left": "",
        "top": "",
        "width": "",
        "height": ""
      },
      "allowCalculateOverride": false,
      "encrypted": false,
      "showCharCount": false,
      "showWordCount": false,
      "allowMultipleMasks": false,
      "leftIcon": "",
      "rightIcon": "",
      "dataGridLabel": true,
      "id": "ef4dg8g",
      "attributes": {},
      "properties": {}
    }
  ]
}
```
Also a jsfiddle link to reproduce
https://jsfiddle.net/rkhfv3gx/2/

**Comment by wag110894** [comment]
Hello,

Thank you for your patience regarding this issue. This has been resolved on the latest version: 4.9.0-beta.4. 

Thank you for using Form.io.
